File: android/app/src/main/java/com/guitaripod/pixie/data/purchases/RevenueCatManager.kt
package com.guitaripod.pixie.data.purchases

import android.app.Activity
import android.app.Application
import android.util.Log
import com.guitaripod.pixie.BuildConfig
import com.revenuecat.purchases.*
import com.revenuecat.purchases.interfaces.*
import com.revenuecat.purchases.models.*
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RevenueCatManager @Inject constructor(
    private val application: Application
) {
    companion object {
        private const val TAG = "RevenueCatManager"
        // RevenueCat Public SDK Key - safe to hardcode
        private const val REVENUECAT_API_KEY = "goog_ekJVbyRpZsNCENTFOjDAzePvssD"
    }
    
    private val _purchaseState = MutableStateFlow<PurchaseState>(PurchaseState.Idle)
    val purchaseState: StateFlow<PurchaseState> = _purchaseState.asStateFlow()
    
    private val _offerings = MutableStateFlow<Offerings?>(null)
    val offerings: StateFlow<Offerings?> = _offerings.asStateFlow()
    
    init {
        setupRevenueCat()
    }
    
    private fun setupRevenueCat() {
        Purchases.debugLogsEnabled = BuildConfig.DEBUG
        
        Purchases.configure(
            PurchasesConfiguration.Builder(application, REVENUECAT_API_KEY)
                .observerMode(false)
                .build()
        )
        
        fetchOfferings()
    }
    
    fun setUserId(userId: String) {
        Purchases.sharedInstance.logIn(
            userId,
            object : LogInCallback {
                override fun onReceived(customerInfo: CustomerInfo, created: Boolean) {
                    Log.d(TAG, "User logged in: $userId, created: $created")
                }
                
                override fun onError(error: PurchasesError) {
                    Log.e(TAG, "Error logging in user: ${error.message}")
                }
            }
        )
    }
    
    fun fetchOfferings() {
        Purchases.sharedInstance.getOfferings(object : ReceiveOfferingsCallback {
            override fun onReceived(offerings: Offerings) {
                _offerings.value = offerings
                Log.d(TAG, "Offerings fetched: ${offerings.all.size} offerings")
            }
            
            override fun onError(error: PurchasesError) {
                Log.e(TAG, "Error fetching offerings: ${error.message}")
                _purchaseState.value = PurchaseState.Error(error.message)
            }
        })
    }
    
    fun purchaseCreditPack(
        activity: Activity,
        packageToPurchase: Package
    ) {
        _purchaseState.value = PurchaseState.Loading
        
        Purchases.sharedInstance.purchase(
            PurchaseParams.Builder(activity, packageToPurchase).build(),
            object : PurchaseCallback {
                override fun onCompleted(
                    storeTransaction: StoreTransaction,
                    customerInfo: CustomerInfo
                ) {
                    val purchaseToken = storeTransaction.purchaseToken
                    val productId = storeTransaction.productIds.firstOrNull() ?: ""
                    
                    _purchaseState.value = PurchaseState.Success(
                        purchaseToken = purchaseToken,
                        productId = productId,
                        packageId = packageToPurchase.identifier
                    )
                    
                    Log.d(TAG, "Purchase successful: $productId")
                }
                
                override fun onError(error: PurchasesError, userCancelled: Boolean) {
                    if (userCancelled) {
                        _purchaseState.value = PurchaseState.Cancelled
                        Log.d(TAG, "Purchase cancelled by user")
                    } else {
                        _purchaseState.value = PurchaseState.Error(error.message)
                        Log.e(TAG, "Purchase error: ${error.message}")
                    }
                }
            }
        )
    }
    
    fun restorePurchases(onComplete: (CustomerInfo?) -> Unit) {
        Purchases.sharedInstance.restorePurchases(object : ReceiveCustomerInfoCallback {
            override fun onReceived(customerInfo: CustomerInfo) {
                Log.d(TAG, "Restore successful")
                onComplete(customerInfo)
            }
            
            override fun onError(error: PurchasesError) {
                Log.e(TAG, "Restore error: ${error.message}")
                onComplete(null)
            }
        })
    }
    
    fun getCustomerInfo(): Flow<CustomerInfo?> = callbackFlow {
        Purchases.sharedInstance.getCustomerInfo(object : ReceiveCustomerInfoCallback {
            override fun onReceived(customerInfo: CustomerInfo) {
                trySend(customerInfo)
            }
            
            override fun onError(error: PurchasesError) {
                Log.e(TAG, "Error getting customer info: ${error.message}")
                trySend(null)
            }
        })
        
        val listener = UpdatedCustomerInfoListener { customerInfo ->
            trySend(customerInfo)
        }
        
        Purchases.sharedInstance.updatedCustomerInfoListener = listener
        
        awaitClose {
            Purchases.sharedInstance.updatedCustomerInfoListener = null
        }
    }
    
    fun resetPurchaseState() {
        _purchaseState.value = PurchaseState.Idle
    }
}

sealed class PurchaseState {
    object Idle : PurchaseState()
    object Loading : PurchaseState()
    data class Success(
        val purchaseToken: String,
        val productId: String,
        val packageId: String
    ) : PurchaseState()
    data class Error(val message: String) : PurchaseState()
    object Cancelled : PurchaseState()
}

File: iOS/Pixie/Pixie/Services/RevenueCatManager.swift
import Foundation
import RevenueCat
import UIKit

enum PurchaseState {
    case idle
    case loading
    case success(purchaseToken: String, productId: String, packageId: String)
    case error(message: String)
    case cancelled
}

@MainActor
final class RevenueCatManager: NSObject {
    static let shared = RevenueCatManager()
    
    private static let revenueCatAPIKey = "appl_GYqkuGCJhGnqPanmoeSJhTvIgHA"
    
    @Published private(set) var purchaseState: PurchaseState = .idle
    @Published private(set) var offerings: Offerings?
    @Published private(set) var customerInfo: CustomerInfo?
    
    private var purchaseStateObservers: [(PurchaseState) -> Void] = []
    
    private override init() {
        super.init()
        setupRevenueCat()
    }
    
    private func setupRevenueCat() {
        Purchases.logLevel = .info
        Purchases.configure(withAPIKey: Self.revenueCatAPIKey)
        Purchases.shared.delegate = self
        fetchOfferings()
    }
    
    func setUserId(_ userId: String) async throws {
        let (customerInfo, created) = try await Purchases.shared.logIn(userId)
        self.customerInfo = customerInfo
        print("User logged in: \(userId), created: \(created)")
    }
    
    func fetchOfferings() {
        Task {
            do {
                let offerings = try await Purchases.shared.offerings()
                self.offerings = offerings
                print("Offerings fetched: \(offerings.all.count) offerings")
            } catch {
                print("Error fetching offerings: \(error.localizedDescription)")
                updatePurchaseState(.error(message: error.localizedDescription))
            }
        }
    }
    
    func purchaseCreditPack(package: Package) async throws {
        updatePurchaseState(.loading)
        
        do {
            let (transaction, customerInfo, userCancelled) = try await Purchases.shared.purchase(package: package)
            
            if userCancelled {
                updatePurchaseState(.cancelled)
                print("Purchase cancelled by user")
            } else {
                let purchaseToken = transaction?.transactionIdentifier ?? ""
                let productId = transaction?.productIdentifier ?? package.storeProduct.productIdentifier
                
                updatePurchaseState(.success(
                    purchaseToken: purchaseToken,
                    productId: productId,
                    packageId: package.identifier
                ))
                
                self.customerInfo = customerInfo
                print("Purchase successful: \(productId)")
            }
        } catch let error as ErrorCode {
            if error == .purchaseCancelledError {
                updatePurchaseState(.cancelled)
                print("Purchase cancelled by user")
            } else {
                updatePurchaseState(.error(message: error.localizedDescription))
                print("Purchase error: \(error.localizedDescription)")
            }
        } catch {
            updatePurchaseState(.error(message: error.localizedDescription))
            print("Purchase error: \(error.localizedDescription)")
        }
    }
    
    func restorePurchases() async throws -> CustomerInfo {
        let customerInfo = try await Purchases.shared.restorePurchases()
        self.customerInfo = customerInfo
        print("Restore successful")
        return customerInfo
    }
    
    func getCustomerInfo() async throws -> CustomerInfo {
        let customerInfo = try await Purchases.shared.customerInfo()
        self.customerInfo = customerInfo
        return customerInfo
    }
    
    func resetPurchaseState() {
        updatePurchaseState(.idle)
    }
    
    func observePurchaseState(_ observer: @escaping (PurchaseState) -> Void) {
        purchaseStateObservers.append(observer)
        observer(purchaseState)
    }
    
    private func updatePurchaseState(_ state: PurchaseState) {
        purchaseState = state
        purchaseStateObservers.forEach { $0(state) }
    }
}

extension RevenueCatManager: PurchasesDelegate {
    nonisolated func purchases(_ purchases: Purchases, receivedUpdated customerInfo: CustomerInfo) {
        Task { @MainActor in
            self.customerInfo = customerInfo
        }
    }
    
    nonisolated func purchases(_ purchases: Purchases, readyForPromotedProduct product: StoreProduct, purchase startPurchase: @escaping StartPurchaseBlock) {
        startPurchase { (transaction, customerInfo, error, userCancelled) in
            print("Promoted product purchase completed")
        }
    }
}